kernel/userspace.c

```diff
@@ -466,8 +466,12 @@ static void unref_check(struct z_object *ko, uintptr_t index)
 	sys_bitfield_clear_bit((mem_addr_t)&ko->perms, index);
 
 #ifdef CONFIG_DYNAMIC_OBJECTS
+	/* TODO: check why this assert hits */
+	/*__ASSERT(IS_PTR_ALIGNED(ko, struct dyn_obj), "unaligned z_object");*/
```

Comment:
@dcpleung @andyross any idea why we are hitting this assert if I uncomment it?

Reply:
You want to be checking IS_PTR_ALIGNED(dyn, struct dyn_obj), no?  It would be pathological, but you could imagine a kernel object aligned on N*8 + 4, with a (8-byte-aligned) dyn_obj field offset 4 or 12 bytes in, and that would be legal.

Beyond that possibility, the question is why this particular kernel object got misaligned.  And as this is a dynamic object, that points to a glitch in the allocator?  Someone should definitely look at this.

Reply:
could be that the object is not allocated using the aligned alloc function?

Reply:
>You want to be checking IS_PTR_ALIGNED(dyn, struct dyn_obj),

It really is the same, because the `struct z_object` is the first member of `struct dyn_obj`. But sure, let me fix that.

Reply:
>could be that the object is not allocated using the aligned alloc function?

@dcpleung I have no idea, could you check?

Reply:
`z_impl_k_object_alloc()` is using `obj_align_get()` to feed alignment to memory allocation, where it should align to `sizeof(void *)`.

Does `IS_PTR_ALIGNED(dyn, struct dyn_obj)` work?